File: Baek_Coding/백준코딩/알고리즘 스터디/숨바꼭질 2.cpp
#define MAX 100001
#include <iostream>
#include <algorithm>
#include <queue>
#include <string.h>

using namespace std;
struct info {
	int mv, idx;
};
info tmp;
int main(){
	int check[MAX];
	int cnt = 0;
	memset(check, 0, sizeof(check));
	int start, end, num,cidx,cmv;
	cin >> start >> end;
	queue<info> q;
	tmp.idx = start;
	tmp.mv = 0;
	q.push(tmp);
	check[start] = 0;
	int dist = 100000;
	while (!cnt) {
		int len = q.size();
		for (int i = 0; i < len; i++) {
			cidx = q.front().idx;
			cmv = q.front().mv;
			check[cidx] = 1;
			q.pop();
			if (cidx == end) { cnt++; dist = min(dist, cmv); }
			if (cidx + 1 < MAX && check[cidx + 1] == 0) {
				tmp.idx = cidx + 1;
				tmp.mv = cmv + 1;
				q.push(tmp);
			}
			if (cidx - 1 >=0 && check[cidx - 1] == 0) {
				tmp.idx = cidx - 1;
				tmp.mv = cmv + 1;
				q.push(tmp);
			}
			if (cidx * 2 < MAX && check[2 *cidx] == 0) {
				tmp.idx = 2 * cidx;
				tmp.mv = cmv + 1;
				q.push(tmp);
			}
		}
	}
	cout << dist << endl << cnt;	
	system("pause");
	return 0;
}

File: Baek_Coding/백준코딩/Practice_Java/숨바꼭질 2.java
import java.util.LinkedList;
import java.util.Queue;
import java.util.Scanner;

class Info {
	int idx, val;

	public Info(int idx, int val) {
		this.idx = idx;
		this.val = val;
	}
}

public class Solution {
	public static void main(String[] args) {
		Queue<Info> q = new LinkedList<>();
		boolean check[] = new boolean[100001];
		Scanner sc = new Scanner(System.in);
		int start = sc.nextInt();
		int end = sc.nextInt();
		Info tmp = new Info(start, 0);
		q.add(tmp);
		int cnt = 0;
		int result=0;
		while (cnt == 0) {
			int len = q.size();
			for (int i = 0; i < len; i++) {
				int cidx = q.peek().idx;
				int cv = q.peek().val;
				q.poll();
				check[cidx]=true;
				if(cidx==end) {
					cnt++;
					result = cv;
					continue;
				}
				if(cidx-1>=0 && !check[cidx-1]) 	q.add(new Info(cidx-1,cv+1));
				if(cidx+1<=100000 && !check[cidx+1]) 	q.add(new Info(cidx+1,cv+1));
				if(2*cidx<=100000 && !check[2*cidx]) 	q.add(new Info(2*cidx,cv+1));
			}
		}
		System.out.print(result +"\n"+cnt);
		sc.close();
	}

}